include/zephyr/net/dsa_core.h

```diff
@@ -0,0 +1,272 @@
+		"swp" STRINGIFY(DT_REG_ADDR(port)), DT_REG_ADDR(port), dsa_port_initialize, NULL,  \
+				&dsa_context_##dsa, &dsa_##dsa##_port_##port##_config,             \
+				CONFIG_ETH_INIT_PRIORITY, &dsa_eth_api, ETHERNET_L2,               \
+				NET_L2_GET_CTX_TYPE(ETHERNET_L2), NET_ETH_MTU);
+
+#define DSA_INIT_INSTANCE(n, _dapi, data)                                                          \
+	struct dsa_context dsa_context_##n = {                                                     \
+		.dapi = _dapi,                                                                     \
+		.prv_data = (void *)data,                                                          \
+		.init_ports = 0,                                                                   \
+		.num_ports = DT_INST_CHILD_NUM_STATUS_OKAY(n),                                     \
+	};                                                                                         \
+	DT_INST_FOREACH_CHILD_STATUS_OKAY_VARGS(n, DSA_PORT_INIT_INSTANCE, n);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** DSA context data */
+struct dsa_context {
+	/** Pointers to all DSA user network interfaces */
+	struct net_if *iface_user[DSA_PORT_MAX_COUNT];
+
+	/** Pointer to DSA conduit network interface */
+	struct net_if *iface_conduit;
+
+	/** DSA specific API callbacks - filled in the switch IC driver */
+	struct dsa_api *dapi;
+
+	/** Number of ports in the DSA switch */
+	uint8_t num_ports;
+
+	/** Number of initialized ports in the DSA switch */
+	uint8_t init_ports;
+
+	/** Instance specific data */
+	void *prv_data;
+};
+
+/**
+ * @brief Structure to provide DSA switch api callbacks - it is an augmented
+ * struct ethernet_api.
+ */
+struct dsa_api {
+	/*
+	 * Callbacks required for DSA switch initialization and configuration.
+	 *
+	 * Each switch instance (e.g. two KSZ8794 ICs) would have its own struct
+	 * dsa_context.
+	 */
+	/** Program (set) mac table entry in the DSA switch */
+	int (*switch_set_mac_table_entry)(const struct device *dev, const uint8_t *mac,
+					  uint8_t fw_port, uint16_t tbl_entry_idx, uint16_t flags);
+
+	/** Read mac table entry from the DSA switch */
+	int (*switch_get_mac_table_entry)(const struct device *dev, uint8_t *buf,
+					  uint16_t tbl_entry_idx);
+
+	/*
+	 * DSA helper callbacks
+	 */
+	/** Handle receive packet for untagging and redirection */
+	struct net_if *(*recv)(struct net_if *iface, struct net_pkt **pkt);
+
+	/** transmit packet with tagging */
+	struct net_pkt *(*xmit)(struct net_if *iface, struct net_pkt *pkt);
+
+	/** port init */
+	int (*port_init)(const struct device *dev);
+
+	/** port link change */
+	void (*port_phylink_change)(const struct device *dev, struct phy_link_state *state,
+				    void *user_data);
+
+	/** switch setup */
+	int (*switch_setup)(const struct dsa_context *dsa_ctx);
+};
+
+enum dsa_port_type {
+	NON_DSA_PORT,
+	DSA_CONDUIT_PORT,
+	DSA_USER_PORT,
+	DSA_CPU_PORT,
+	DSA_PORT,
```

Comment:
Doxygen documentation missing for these.

Reply:
I should have moved them out of INTERNAL_HIDDEN.
Will do that.
Thanks.

Reply:
If these are not to be documented, then the `/** @cond INTERNAL_HIDDEN */` line below should be moved before this enum, and then the previous hidden doc block can be combined with this one.
But if you meant that these are to be documented, then the documentation is still missing.
I did not quite understand your comment here.

I think the `/** @cond INTERNAL_HIDDEN */` blocks are incorrectly set now. You have documentation of the APIs but they are marked now as hidden which prevents documentation generation. Please readjust these checks.

Reply:
Sorry, I just realized it's my code style problem.
Going through the header file, the only thing needs to public is one API `dsa_user_get_iface` used in sample.
Others were used only in drivers. Will fix issues.
Thank you very much.